sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala

```diff
@@ -278,4 +284,35 @@ class QueryExecutionErrorsSuite extends QueryTest
     assert(e.getMessage ===
       "Datetime operation overflow: add 1000000 YEAR to TIMESTAMP '2022-03-09 01:02:03'.")
   }
+
+  test("CANNOT_PARSE_DECIMAL: unparseable decimal") {
+    val e1 = intercept[SparkException] {
+      val schema = new StructType().add("money", DecimalType.DoubleDecimal)
+      spark
+        .read
+        .schema(schema)
+        .format("csv")
+        .option("header", "true")
+        .option("locale", Locale.ROOT.toLanguageTag)
+        .option("multiLine", "true")
+        .option("inferSchema", "false")
+        .option("mode", "FAILFAST")
```

Comment:
Just wonder, is it possible to use the `from_csv` function? If so, we could avoid access to the file system.

Reply:
### from_csv don't support FailFastMode, code as folllow:
if (mode != PermissiveMode && mode != FailFastMode) {
      throw QueryCompilationErrors.parseModeUnsupportedError("from_csv", mode)
}

### In FailureSafeParser:
def parse(input: IN): Iterator[InternalRow] = {
    try {
      rawParser.apply(input).iterator.map(row => toResultRow(Some(row), () => null))
    } catch {
      case e: BadRecordException => mode match {
        case PermissiveMode =>
          Iterator(toResultRow(e.partialResult(), e.record))
        case DropMalformedMode =>
          Iterator.empty
        case FailFastMode =>
          throw QueryExecutionErrors.malformedRecordsDetectedInRecordParsingError(e)
      }
    }
  }
if Mode is not FailFastMode, the deep exception: CANNOT_PARSE_DECIMAL, can't throw。